### PR TITLE
#3: Provide PasswordResetFlow (closes #3)

### LINKED ISSUE #3: Provide PasswordResetFlow
Based on PasswordResetAuthenticationDomain and LoginAuthenticationDomain

### PR DESCRIPTION
**Pull request by bytecodeguru**
Closes #3

New flow for token based authentication recovery, with two basic steps and a twist:

- `registerForRecovery`, to generate a new recovery token that can be sent to the specified subject, giving access to a reset password form.

- `recoverLogin`, to complete the recovery flow, consuming the recovery token by specifying a new login password.

The twist is that, by design, there's no mean for toctoc to know the right username that has to be associated to the new password. So, for this first iteration, we are exposing a callback that the application can use to provide the right username.

This could change in the future, because toctoc should support also use cases where a user can be associated to more than one login credentials.

## Test Plan

~Nothing to see here, maybe after https://github.com/buildo/backend/issues/17~

Property based tests with InMemoryAuthenticationDomain test doubles 💪

### CODE REVIEW

**Comment by gabro** [comment]
@bytecodeguru tests are failing

File: toctoc/core/src/main/scala/io.buildo.toctoc/core/authentication/TokenBasedRecovery.scala [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+  class TokenBasedRecoveryFlow[F[_]: Monad](
+    loginD: LoginAuthenticationDomain[F],
+    recoveryTokenD: AccessTokenAuthenticationDomain[F],
+    tokenDuration: Duration = Duration.ofDays(365),
+  ) extends BCryptHashing {
+    def registerForRecovery(s: Subject): F[Either[AuthenticationError, AccessToken]] =
+      (for {
+        rtd <- EitherT(recoveryTokenD.unregister(s))
+        token = AccessToken(randomString(64), Instant.now().plus(tokenDuration))
```
Comment (gabro) [inline toctoc/core/src/main/scala/io.buildo.toctoc/core/authentication/TokenBasedRecovery.scala]: maybe we can add a constructor to `AccessToken` that uses these defaults:

```scala
object AccessToken {
  def generate(duration: Duration): AccessToken = AccessToken(randomString(64), Instant.now.plus(duration)
}
```

and then

```suggestion
        token = AccessToken.generate(tokenDuration)
```

File: toctoc/slickPostgreSql/src/main/scala/io.buildo.toctoc/slick/authentication/PostgreSqlSlickTokenBasedRecoveryFlow.scala [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+import login.PostgreSqlSlickLoginAuthenticationDomain
+
+import monix.catnap.FutureLift
+import cats.effect.Sync
+import _root_.slick.jdbc.JdbcBackend.Database
+
+import scala.concurrent.Future
+import java.time.Duration
+
+class PostgreSqlSlickTokenBasedRecoveryFlow[F[_]: Sync: FutureLift[?[_], Future]](
```
Comment (gabro) [inline toctoc/slickPostgreSql/src/main/scala/io.buildo.toctoc/slick/authentication/PostgreSqlSlickTokenBasedRecoveryFlow.scala]: we've decided to remove the specific flows from toctoc, so I wouldn't add this one.

File: toctoc/core/src/main/scala/io.buildo.toctoc/core/authentication/TokenBasedRecovery.scala (hunk 1 above)
Comment (bytecodeguru) [reply toctoc/core/src/main/scala/io.buildo.toctoc/core/authentication/TokenBasedRecovery.scala]: Nice idea, I'd leave `randomString` out of scope and open new issues for:
- `randomString` comes from `BCryptHashing`
- `Instant.now` could be handled in a more generic way, perhaps something like `Clock[F[_]]` from cats-effect.

Comment (gabro) [reply toctoc/core/src/main/scala/io.buildo.toctoc/core/authentication/TokenBasedRecovery.scala]: Sounds good. Let's open new issues

Comment (bytecodeguru) [reply toctoc/core/src/main/scala/io.buildo.toctoc/core/authentication/TokenBasedRecovery.scala]: done ✅

File: toctoc/core/src/main/scala/io.buildo.toctoc/core/authentication/TokenBasedRecovery.scala [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+import cats.Monad
+import cats.data.EitherT
+
+import java.time.Duration
+
+object TokenBasedRecovery {
+  import TokenBasedAuthentication._
+
+  final case class TokenBasedRecoveryFlow[F[_]: Monad] private (
```
Comment (gabro) [inline toctoc/core/src/main/scala/io.buildo.toctoc/core/authentication/TokenBasedRecovery.scala]: Two doubts:
- why a case class?
- what's the purpose of the `object TokenBasedRecovery`? Namespacing?

File: toctoc/core/src/test/scala/io.buildo.toctoc/core/authentication/TokenBasedRecoveryFlowSpec.scala [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+  val loginDomain: LoginDomain[IO] =
+    InMemoryAuthenticationDomain.create(Map.empty)
+
+  val tokenDomain: AccessTokenDomain[IO] =
+    InMemoryAuthenticationDomain.create(Map.empty)
+
+  val recoveryFlow = TokenBasedRecoveryFlow.create(loginDomain, tokenDomain, Duration.ofDays(1))
+
+  def check[A](test: => IO[Either[A, Assertion]]): Assertion =
```
Comment (gabro) [inline toctoc/core/src/test/scala/io.buildo.toctoc/core/authentication/TokenBasedRecoveryFlowSpec.scala]: I like the idea of making this a utility 👍

File: toctoc/core/src/test/scala/io.buildo.toctoc/core/authentication/TokenBasedRecoveryFlowSpec.scala [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    forAll { (s: UserSubject, username: String, password: String) =>
+      check {
+        (for {
+          registerResult <- EitherT(recoveryFlow.registerForRecovery(s))
+          (f, token) = registerResult
+          f2 <- EitherT(f.recoverLogin(token, password) { _ =>
+            IO(Some(username))
+          })
+          authenticateResult <- EitherT(f2.loginD.authenticate(Login(username, password)))
+          (_, subject) = authenticateResult
```
Comment (gabro) [inline toctoc/core/src/test/scala/io.buildo.toctoc/core/authentication/TokenBasedRecoveryFlowSpec.scala]: super minor, but I kinda hate that we cannot do:

```scala
(_, subject) <- EitherT(...)
```

This is solved by https://github.com/oleg-py/better-monadic-for which we might want to introduce in our build (I've used in on personal projects and it works well!)

File: toctoc/core/src/test/scala/io.buildo.toctoc/core/authentication/TokenBasedRecoveryFlowSpec.scala [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+      check {
+        (for {
+          registerResult <- EitherT(recoveryFlow.registerForRecovery(s))
+          (f, token) = registerResult
+          result <- EitherT(f.recoverLogin(token, password) { _ =>
+            IO(None)
+          })
+        } yield {
+          result
+        }).swap.map {
```
Comment (gabro) [inline toctoc/core/src/test/scala/io.buildo.toctoc/core/authentication/TokenBasedRecoveryFlowSpec.scala]: why not `leftMap`?

**Pull request merged**